10k PRs in a month? easy! (#18790)
PostHog/posthog.com: contents/blog/10k-prs-in-a-month.md

---
date: "2026-07-22"
title: "10k PRs in a month? easy!"
author:
  - paul-dambra
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
category: Engineering
tags:
  - AI
  - Engineering
---

> ## AI pilled

Author: ```suggestion

The north star for internal developer experience at PostHog is "10,000 PRs in a month is easy". As agent workflows accelerate what is possible. We're sharing what we're changing and how we've been able to increase the volume of engineering work and keep delivering reliable systems

## AI pilled
```

---
date: "2026-07-22"
title: "10k PRs in a month? easy!"
author:
  - paul-dambra
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
category: Engineering
tags:
  - AI
  - Engineering
---

The north star for internal developer experience at PostHog is "10,000 PRs in a month is easy". As agent workflows accelerate what is possible. We're sharing what we're changing and how we've been able to increase the volume of engineering work and keep delivering reliable systems

## AI pilled

*Covers:*

- No token budget — trust and feedback over process
- Individuals building their own review and PR management tools
- Folk at the edges: 20–40 PRs open, agentic flows managing them
- 20% → 70% of monorepo PRs with agent involvement in 4 months; self-driving ~4% and growing fast

> The stuff I think we should cover, there's no token budget at PostHog. We want people to use the right amount of tokens for their work. Actually, why would we cover that? We would cover that because we want to talk about how the world of agentic programming and leaning into that has changed how we think about and how we do developer experience at PostHog. The most important thing being that we're not putting a budget in front of people. In all the things we do at PostHog, we value trust and feedback over process. So if we find we get to the point where we're uncomfortable with someone's spend, we'll dig into that with them and figure out if the spend is reasonable.

Author: ```suggestion
There's no token budget at PostHog. We can see that the world of product engineering is changing rapidly. And we want people to figure out what that means for us... how we make the best use of the new tools and create new tools with them. So we want people to use the right amount of tokens for their work. And we're not going to pretend we know what that amount is right now. 

The most important thing being that we're not putting a budget in front of people. In all the things we do at PostHog, we value trust and feedback over process. So if we find we get to the point where we're uncomfortable with someone's spend, we'll dig into that with them and figure out if the spend is reasonable. And what we can learn from that for everyone else.
```

## AI pilled

*Covers:*

- No token budget — trust and feedback over process
- Individuals building their own review and PR management tools
- Folk at the edges: 20–40 PRs open, agentic flows managing them
- 20% → 70% of monorepo PRs with agent involvement in 4 months; self-driving ~4% and growing fast

There's no token budget at PostHog. We can see that the world of product engineering is changing rapidly. And we want people to figure out what that means for us... how we make the best use of the new tools and create new tools with them. So we want people to use the right amount of tokens for their work. And we're not going to pretend we know what that amount is right now. 

The most important thing being that we're not putting a budget in front of people. In all the things we do at PostHog, we value trust and feedback over process. So if we find we get to the point where we're uncomfortable with someone's spend, we'll dig into that with them and figure out if the spend is reasonable. And what we can learn from that for everyone else.

> At the individual level, we have people writing their own review tools and their own PR management tools because we've found that being able to speed up the rate at which we can create PR's is not a company by being able to speed up the rate at which we can deal with PR's. So building and testing and validating in production still have to happen even where we can speed up the work that we're doing.

Author: ```suggestion
At the individual level, we have people writing their own review tools and their own PR management tools. We've found that being able to speed up the rate at which we can create PR's doesn't mean we've been able to speed up the rate at which we can deal with PR's. So building and testing and validating in production still have to happen even where we can speed up the work that we're doing.
```

> and that's not all we're doing at dev x at posthog but those are the really top level things then you see the folk who are out at their edges really pushing the limits of what they can achieve have written their own applications they have you know the 20 30 40 prs open at a time they have agentic flows keeping those prs up to date with trunk making sure that ci is passing making sure that reviews have been requested and that's all stuff that we're not we're happy for folk to spend the time learning that and it means we can fold it back into either what we offer to everyone at PostHog internally or what we offer to our customers as we let their products build themselves

Author: ```suggestion
At the individual level, we have people writing their own review tools and their own PR management tools. We've found that being able to speed up the rate at which we can create PR's doesn't mean we've been able to speed up the rate at which we can deal with PR's. 

The folk who are really pushing the limits of what they can achieve have written their own applications. They have maybe 40 prs open at the same time. The only way to make that work is to have agentic loops getting those PRs ready for human attention. Things like keeping up to date with trunk, making sure that ci is passing, making sure that reviews have been acted on
```

> Over the last 4 months we've moved from around 20% of our monorepo PRs being opened with agent involvement to 70% (that's not PRs with no human involvement :)). Self-driving  product is around 4% of our PRs right now but growing very fast.

Author: ```suggestion
$$$ something about skills and scouts $$$

Over the last 4 months we've moved from around 20% of our monorepo PRs being opened with agent involvement to 70% (that's not PRs with no human involvement :)). Self-driving  product is around 4% of our PRs right now but growing very fast.
```

## Local dev

*Covers:*

- Heavy stack vs. agents wanting multiple copies running — macbooks can't keep up
- Skill-driven dev machines: ask your agent to start a dev box, test, record a video
- Cloud runners — for our stack and anybody's
- hogli custom CLI; pushing CI down into local development

> We've been working on local dev machines the PostHog stack is very heavy that's always meant that we've needed very powerful dev machines and that was fine up until the point now where people want to run multiple copies of the system test multiple prs that they're working on if they're being really leaning into agentic development and there's only so much ram and disk you can put into a macbook so we're building out literal dev machines they're skill driven so you just ask your agent to start a dev machine and test the work that you're doing quite frequently instead of driving the site myself i just tell an agent session start a dev box, run this copy of the site in it, record me a video of the changes that we're making so i can see the output of the test rather than do the test myself in PostHog desktop we're also building cloud runners so we're making a way for cloud runners to run not just our stack but anybody's stack so you can do the same type of work in there so for local dev the reaction is mostly no limits on tokens and strong local dev flow we have a custom cli we call hog li or hogli which does gives us a way to help people out so we've pushed a bunch of ci down into local development

Author: ```suggestion
The PostHog stack is very heavy for local development. that's always meant that we've needed very powerful dev machines and that was fine up until the point where people want to run multiple copies of the system to test multiple prs their agents are working on. and there's only so much ram and disk you can put into a macbook.

 so we're building out literal dev machines. they're skill driven so you just ask your agent to start a dev machine and test the work that you're doing. a prompt like: "start a dev box, run this PR in it, record me a video of the changes that we're making so i can see the output of the test"

we also have a custom cli we call hogli. it advertises a tool so that agents send feedback to our devex team when something is difficult for them. And gives us an easy way to push a bunch of our CI checks down into local development machines.
```

---
date: "2026-07-22"
title: "10k PRs in a month? easy!"
author:
  - paul-dambra
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
category: Engineering
tags:
  - AI
  - Engineering
---

> The north star for internal developer experience at PostHog is "10,000 PRs in a month is easy". As agent workflows accelerate what is possible. We're sharing what we're changing and how we've been able to increase the volume of engineering work and keep delivering reliable systems

Reviewer: i think a "why" here would help, like "we're building for 10K/month so we can support our dev team at any scale" or w/e

Reviewer 2: 100%... this goes directly to the most likely response from negative folks: "10,000 slop PRs is a meaningless metric". and I agree we should not chase it for chasing's sake

Author: ```suggestion
The north star for internal developer experience at PostHog is "10,000 PRs in a month is easy". As agent workflows accelerate what is possible. We're sharing what we're changing and how we've been able to increase the volume of engineering work and keep delivering reliable systems. That's not because we think that more PRs is a good thing in and of itself, but because we hire talented, driven people and we're investing in agentic automation. That means that the number of PRs will climb.

If it doesn't... well, then we've made local dev, CI, and validation faster and better. That's a great outcome whatever the reason driving it
```

Author: how about that?

Reviewer: love it 👍🏼

At the individual level, we have people writing their own review tools and their own PR management tools. We've found that being able to speed up the rate at which we can create PR's doesn't mean we've been able to speed up the rate at which we can deal with PR's. 

The folk who are really pushing the limits of what they can achieve have written their own applications. They have maybe 40 prs open at the same time. The only way to make that work is to have agentic loops getting those PRs ready for human attention. Things like keeping up to date with trunk, making sure that ci is passing, making sure that reviews have been acted on

$$$ something about skills and scouts $$$

Over the last 4 months we've moved from around 20% of our monorepo PRs being opened with agent involvement to 70% (that's not PRs with no human involvement :)). Self-driving  product is around 4% of our PRs right now but growing very fast.

> ## Local dev

Author: @MattPua   
  

> would love to see something more on the actual numbers of PRs open at any given time/per user (not power users per say but average), and how we've changed processes to have that happen (beyond just token spend, like slack bot, etc)

i guess that's a "dev" thing rather than a "ci" thing  
  
i don't imagine we can get historic data for how many open PRs each of us have i can measure it today, and we can ask folk if it has changed for them  
  
why is that a thing to comment on?  
the cognitive load of balancing multiple things?  
  
what do you think is making it work or not?  
  
i actually think this is interesting enough to be a follow-up piece!

Reviewer 3: i think the more interesting story IMO (CI story is great, dont get me wrong), is the way we are perceiving pRs and how many people are opening and juggling how many tasks at a given time. I think if you asked this a year ago, people might have had one or two PRs at a time, and that was all that they could fit in their context window of their heads. That made sense. 

I think for us, because we are trying to be as AI-pilled as possible, we're trying to go really far ahead with the concept of how many PRs is too many, almost. What is that limit? What do we as PostHog do to enable people to even write and open that many PRs to begin with? If you think about it, we're talking about something like more than 20 PRs a person a month. Probably for a lot of orgs, that's a crazy number to think about. Probably what they would care about is: how does that even happen, or what can they do to enable their orgs to make that happen? 

I think a lot of teams, from the CI perspective, already are kind of experiencing this problem: how does CI handle the workloads of this many PRs? GitHub is already a prime example ofthat. But I think you have a lot fewer teams, at least engineering teams, that are talking about how they make that culture happen. I think that's going to be the more interesting and unique thing we could talk about that's very PostHog.

Author: i used to start to worry at around 15 open PRs (have always multitasked)  
  
and now i start to worry at about 40

*Covers:*

- Heavy stack vs. agents wanting multiple copies running — macbooks can't keep up
- Skill-driven dev machines: ask your agent to start a dev box, test, record a video
- Cloud runners — for our stack and anybody's
- hogli custom CLI; pushing CI down into local development

The PostHog stack is very heavy for local development. that's always meant that we've needed very powerful dev machines and that was fine up until the point where people want to run multiple copies of the system to test multiple prs their agents are working on. and there's only so much ram and disk you can put into a macbook.

>  so we're building out literal dev machines. they're skill driven so you just ask your agent to start a dev machine and test the work that you're doing. a prompt like: "start a dev box, run this PR in it, record me a video of the changes that we're making so i can see the output of the test"

Reviewer 2: can also mention we tried github codespaces in the past but saw even less adoption; we have to meet developers where they are more and more for any new tooling introduced

Author: "we have to meet developers where they are more and more for any new tooling introduced" love this

> we also have a custom cli we call hogli. it advertises a tool so that agents send feedback to our devex team when something is difficult for them. And gives us an easy way to push a bunch of our CI checks down into local development machines.

Reviewer 2: btw one neat trick that we (devex) have used to steer agentic behavior is... enforce agents read the PR template before creating a PR and then adding instructions there. not sure if this fits somewhere in local dev section but wanted to mention it as it powers auto-mermaid diagrams, PR screenshots etc

Then we can look at PR titles, since we use conventional commits to signal the type of work in the PR.

| Prefix | May 2025 (count / share) | Nov 2025 | Jun 2026 |
|---|---|---|---|
| Total PRs/mo (all PRs, not just merged) | 755 | 1,126 | 4,868 |
| fix | 309 / 40.9% | 430 / 38.2% | 1,997 / 41.0% |
| feat | 259 / 34.3% | 397 / 35.3% | 1,782 / 36.6% |
| chore | 154 / 20.4% | 250 / 22.2% | 823 / 16.9% |
| refactor | 15 / 2.0% | 28 / 2.5% | 130 / 2.7% |
| perf | 4 / 0.5% | 5 / 0.4% | 68 / 1.4% |
| revert | 1 / 0.1% | 5 / 0.4% | 9 / 0.2% |

We see that composition is very stable across the last year or so. Importantly reverts as a proportion of PRs is very stable. And we are super eager to revert whenever we see a need to. So, that's a good signal that production is not breaking more under the load of agentic PRs

An aside that reliability of PostHog overall is going up too, $$$some reliability fact here$$$

flakey test suites are the biggest problem we face here. $$$ and so we've done $$$
## Review

*Covers:*

- StampHog: $300/month in tokens, auto-approves ~20% of PRs based on safety + familiarity
- ReviewHog: fleet-of-agents thorough review for anything not simple

> - Humans review way less, reviews way more effective — focus on understanding, safety, value

Reviewer 2: >Humans review way less

i am not sure this is true; according to https://us.posthog.com/project/2/dashboard/1860936

Another useful dashboard fyi https://us.posthog.com/project/2/dashboard/1812721

Author: As in the trajectory will be that as PRs grow humans have to review less, and we have to figure out how to make that safe

Author: ```suggestion
- As PRs arrive faster and faster humans simply have to review less, and we have to figure out how to make that safe
```

---
date: "2026-07-24"
title: "10,000 PRs a month is easy: How devex is evolving at PostHog"
author:
  - paul-dambra
rootPage: /blog
sidebar: Blog
featuredImage: >-
  https://res.cloudinary.com/dmukukwp6/image/upload/v1783662227/stop_being_the_code_review_bottleneck_hero_a734a37558.png
featuredImageType: full
showTitle: true
hideAnchor: true
category: Engineering
tags:
  - AI
  - Engineering
---

> Shipping cadence is accelerating at PostHog. In the last 6 months, we've gone from shipping 1,441 PRs in January to 4,725 PRs in June while increasing engineering headcount only 10%.

Reviewer 4: ```suggestion
Shipping cadence is accelerating at PostHog. In the last 6 months, we've gone from shipping 1,441 PRs in January to 4,725 PRs in June while increasing engineering headcount by only 10%.
```

![More PRs](https://res.cloudinary.com/dmukukwp6/image/upload/w_1600,c_limit,q_auto,f_auto/Clean_Shot_2026_07_23_at_15_27_38_2x_1a2d8df53b.png)

> This is heading towards our north star for internal developer experience: "10,000 PRs a month is easy." We thought we'd have until December, but this might come as soon as September.

Reviewer 4: ```suggestion
This is heading towards our north star for internal developer experience: "10,000 PRs a month is easy." We thought we would have until December to hit this number, but we might get there as early as September.
```

> We don't think more PRs is a good thing in and of itself. It is a consequence of hiring talented, driven people and investing in agentic automation. Even if we don't hit these numbers, devex work still makes local dev, CI, and validation faster and better.

Reviewer 4: ```suggestion
More PRs isn't a good thing in and of itself. It happens as a result of hiring talented, driven people and investing in agentic automation. Even if we don't hit these numbers, working toward our goals means making local dev, CI, and validation faster and better.
```

> We're sharing how we've been able to increase the volume of engineering work while still delivering reliable systems, and the changes we're making to make sure this remains the case.

Reviewer 4: ```suggestion
Here's how we've increased the volume of engineering work while still delivering reliable systems, and the changes we're making to make sure this remains the case.
```

## What being AI-pilled looks like

> There's no token budget at PostHog. We want people to figure out how we _make use of new tools_ and _create new tools_. It's up to them to use the right amount of tokens and we're not pretending we know what that amount is. 

Reviewer 4: ```suggestion
There's no token budget set by some higher-ups at PostHog. We want people to figure out how we _make use of new tools_ and _create new tools_. It's up to them as individuals to use the right amount of tokens, and none of us are going to pretend we know what that amount is. 
```

Reviewer 4: helps set the context that this is an eng leadership post

> At PostHog, we value trust and feedback over process. If we get to the point where we're uncomfortable with someone's spend, we'll dig into that with them and figure out if the spend is reasonable. Everyone else can learn from that too.

Reviewer 4: ```suggestion
This works because it's a big part of our culture to value trust and feedback over process. If we get to the point where we're uncomfortable with someone's spend, we'll dig into that with them and figure out if the spend is reasonable. Everyone else can learn from that too.
```

Speeding up the rate we create PRs doesn't necessarily mean speeding up the rate we can handle them. For that, we're relying on agentic loops getting those PRs ready for human attention. Things like: 

- Keeping up to date with trunk
- Making sure that CI is passing
- Making sure that reviews have been acted on

The folk who are pushing the limits of what they can achieve have written their own applications, like review and PR management tools or [`qa-swarm`](https://github.com/pauldambra/dotfiles/tree/main/ai/skills/qa-swarm) and [`babysit-prs`](https://github.com/haacked/dotfiles/blob/main/ai/skills/babysit-prs/SKILL.md) skills. They have maybe 40 PRs open at the same time, but we want to push this further and get more people onboard. Something like 60% of my token spend is burned automating the toil of handling CI and review and I don't regret a single dollar.

## Getting off local machines

> The PostHog stack is very heavy for local development. That's always meant that we've needed powerful dev machines. This is fine until people want to run multiple copies test PRs. We're hitting the limits of what's possible; there's only so much RAM and disk you can put into a Macbook.

Reviewer 4: ```suggestion
The PostHog stack is very heavy for local development, so we've always needed powerful dev machines. This was fine until people started wanting to run multiple copies of the stack to test PRs. We're hitting the limits of what's possible now; there's only so much RAM and disk you can put into a Macbook.
```

> So we're building out cloud dev machines. They're skill-driven so you just ask your agent to start a dev machine with a prompt like: 

Reviewer 4: ```suggestion
So, we're building out cloud dev machines. They're skill-driven so you just ask your agent to start a dev machine with a prompt like: 
```

> > Start a dev box, run this PR in it, and record me a video of the changes that we're making so I can see the output of the test

Reviewer 4: ```suggestion
> Start a dev box, run this PR in it, and record me a video of the changes that we're making so I can see the output of the test.
```

> Powering this is our [internal developer CLI `hogli`](https://github.com/PostHog/posthog/tree/master/tools/hogli). Along with managing cloud dev machines, it pushes CI checks down into local dev by helping agents better test code locally. It also includes a tool to send feedback to our devex team when something is difficult (for people or agents), helping our devex team make specific improvements fast.

Reviewer 4: ```suggestion
This is powered by our [internal developer CLI `hogli`](https://github.com/PostHog/posthog/tree/master/tools/hogli), which also pushes CI checks down into local dev by helping agents better test code locally. It also includes a tool to quickly send feedback to our devex team when something is difficult (for people or agents), which helps our devex team make specific improvements fast.
```

## Battling the bottleneck of CI

> Like many companies, CI has always been a challenge at PostHog. We've always had a big CI suite due to the breadth of our product. Increasing PRs created 50% month-over-month is like spraying a house fire with gasoline. 

Reviewer 4: ```suggestion
Like many companies, CI has always been a challenge at PostHog because of the breadth of our product. That means increasing the number of PRs created by 50% was like spraying a house fire with gasoline. 
```

> [Depot](https://depot.dev/customers/posthog), the compute infrastructure for our CI, has handled this incredibly well given that in June 2026, we used 21,643,653 minutes of CI time. That's 41 _years_ of CI time in June.

Reviewer 4: ```suggestion
[Depot](https://depot.dev/customers/posthog), the compute infrastructure for our CI, has handled this incredibly well – in June 2026, we used 21,643,653 minutes of CI time. That's 41 _years_ of CI time in a single month.
```

> The key part of making our CI more efficient is the work **not** done. Basically, we want to skip the CI more, but not too much. We need it to be in the Goldilocks zone. This means working on:

Reviewer 4: ```suggestion
The key to making our CI more efficient is defining the work that does **not** get done. Basically, we want to skip the CI more, but not too much; we need it to be in the Goldilocks zone. This means working on:
```

- Ways of quarantining flaky tests. We can mark tests so they run and report, but can't fail the CI, blocking everyone.

![Quarantining](https://res.cloudinary.com/dmukukwp6/image/upload/w_1600,c_limit,q_auto,f_auto/Clean_Shot_2026_07_24_at_10_46_19_2x_75584c1e76.png)

> - Building out the abstract syntax tree (AST) for changes so we can more selectively run tests. Right now, we have too many PRs where the whole test suite is running unnecessarily because the CI can't tell a change is contained from the static analysis of paths.

Reviewer 4: ```suggestion
- Building out the abstract syntax tree (AST) for changes so we can more selectively run tests. Right now, we have too many PRs where the whole test suite is running unnecessarily because the CI can't tell when a change is contained from the static analysis of paths.
```

> - Merge queues as we're increasingly seeing passing PRs break main when merged together. 

Reviewer 4: ```suggestion
- Merge queues, since we're increasingly seeing passing PRs break main when merged together. 
```

- Increasing the speed of CI jobs like using a pre-migrated database schema (which decreased backend CI times from ~23 to ~15 minutes). 

> Alongside all this, we're streaming [GitHub data](/docs/data-warehouse/sources/github) into PostHog. This lets us have engineering analytics set up for ourselves and look at it with [PostHog AI](/ai):

Reviewer 4: ```suggestion
Alongside all this, we're streaming [GitHub data](/docs/data-warehouse/sources/github) into PostHog. This gives us engineering analytics that we can look at using [PostHog AI](/ai):
```

- What workflow jobs cost the most money?
- Which take the most time?
- Which fail most frequently?

This lets us make targeted improvements on the CI and get PRs to green faster.

> ## How we validate it isn't breaking

Reviewer 4: ```suggestion
## How we know it isn't breaking
```

> We've always been more interested in being able to quickly fix things over avoiding breaking things in the first place. In areas like UI, charts, access, and filters, it's much more important that we find out if the change is useful than getting it perfect first time. 

Reviewer 4: ```suggestion
We've always cared more about making it easy to quickly fix things rather than avoiding breaking them in the first place. This is especially true when dealing with UI, charts, access, and filters. It's much more important that we find out if the change is actually useful instead of wasting time getting it perfect on the first try.
```

> But we know there's a lot of stuff that can't break. Accepting and displaying your data is the core use case of PostHog for many users. Failing at this is catastrophic, so we rely on a lot of automated testing and validation. We've written a quarter of a million Python tests for our backend. They're paying off even more now because they mean we have higher confidence in agent driven work where validation says it's safe.

Reviewer 4: ```suggestion
But in reality, there's a lot of stuff that we can't let break. For example, accepting and displaying your data – it's the core reason many users come to PostHog. Failing at this would be catastrophic, so we rely on a lot of automated testing and validation, including a quarter of a million Python tests for our backend. They're paying off even more now because they give us higher confidence in agent-driven work where validation says it's safe.
```

Safety is not quality though, so we have some measures to help us see how things are changing. 

> The first is size. Line count is a rough proxy. Past some threshold people stop reviewing carefully and quality drops. Agents are writing bigger PRs but not hugely bigger. Partly that is our engineers driving quality.

Reviewer 4: i don't quite understand this line. is it saying "longer PRs have lower quality"? and so with agents writing bigger PRs, we are dropping in quality?

| Percentile | Wk of May 25 (lines) | Wk of Jul 13 (lines) | Change |
| ---------- | -------------------- | -------------------- | ------ |
| p50        | 120                  | 144                  | +20%   |
| p75        | 398                  | 760                  | +91%   |
| p90        | 1,004                | 1,432                | +43%   |
| p99        | 7,718                | 14,023               | +82%   |

> Another way is PR titles since we use conventional commits like `fix`, `feat`, and `revert` to signal the type of work in the PR.

Reviewer 4: ```suggestion
Another useful mechanism is PR title conventions. We use conventional commits like `fix`, `feat`, and `revert` to signal the type of work in the PR.
```

| Prefix              | May 2025 (count/share) | Nov 2025    | Jun 2026      |
| ------------------- | ---------------------- | ----------- | ------------- |
| Total merged PRs/mo | 755                    | 1,126       | 4,868         |
| fix                 | 309 / 40.9%            | 430 / 38.2% | 1,997 / 41.0% |
| feat                | 259 / 34.3%            | 397 / 35.3% | 1,782 / 36.6% |
| chore               | 154 / 20.4%            | 250 / 22.2% | 823 / 16.9%   |
| refactor            | 15 / 2.0%              | 28 / 2.5%   | 130 / 2.7%    |
| perf                | 4 / 0.5%               | 5 / 0.4%    | 68 / 1.4%     |
| revert              | 1 / 0.1%               | 5 / 0.4%    | 9 / 0.2%      |

> The composition is stable across the last year or so. Importantly reverts as a proportion of PRs has barely changed. And we are eager to revert whenever we see a need to. So, that's a good signal that production is not breaking more under the load of agentic PRs.

Reviewer 4: ```suggestion
The distribution of PR types is pretty stable across the last year or so. An important callout, though, is that the percentage of `reverts` has barely changed. We're eager to revert whenever we see a need to, so this is a good sign that implies that production is _not_ breaking more under the load of agentic PRs.
```

> The ultimate goal is reliability for all of PostHog which has remained solid. 

Reviewer 4: ```suggestion
The ultimate goal is reliability for all of PostHog, which has remained solid:
```

| Service                      | Uptime (90d, Apr 26–Jul 24 2026) |
| ---------------------------- | -------------------------------- |
| Event ingestion (success)    | 100.00%                          |
| Session replay ingestion     | 100.00%                          |
| Feature Flags API            | 100.00%                          |
| App                          | 99.993%                          |
| REST API query endpoints     | 99.985%                          |
| All other REST API endpoints | 99.951%                          |
| PostHog AI                   | 99.927%                          |

## Humans don't need to review every PR

With agents creating more and more of the PRs, [code review](/newsletter/code-review-tips) becomes a bigger part of the work. Engineers spend less time on each review, so we need to find ways of making that safe while maintaining momentum.

> We require approvals on all our PRs, but an increasing number of those approvals are coming from StampHog. It runs deterministic safety checks then reviews with an LLM for showstoppers. If it's happy, it approves. If not, it escalates to human review. 

Reviewer 4: ```suggestion
We require approvals on all our PRs, but an increasing number of those approvals are coming from StampHog, an agent that runs deterministic safety checks then reviews with an LLM for showstoppers. If it's happy, it approves. If not, it escalates to human review. 
```

> We put it together by looking at:

Reviewer 4: ```suggestion
We built StampHog by looking at:
```

- The open history of PRs in PostHog and asking "what does a PR look like when it's approved without any further changes?" 
- How safe the area of the application is to change 
- How familiar with that part of the system the engineer opening the PR is

20% of our PRs are approved by StampHog and it only costs us about $300 per month in tokens. This is an incredibly efficient investment for the speed up it gives engineers who can now deliver small safe steps very quickly. 

For more complicated reviews, we're also building out ReviewHog which we'll eventually offer as part of self-driving product. It gives us the top-down, thorough review by a fleet of agents.

> These let engineers concentrate on the genuinely important stuff: 

Reviewer 4: ```suggestion
Combined, these agentic tools let our engineers concentrate on the genuinely important questions that need human attention, like: 
```

> - Do they understand this change?

Reviewer 4: ```suggestion
- "Do I understand this change?"
```

Reviewer 4: i think quotes are necessary here to imply it's a question being asked by the dev as they review!

> - Does the author understand the change?

Reviewer 4: ```suggestion
- "Does the author understand the change?"
```

> - Is it safe?

Reviewer 4: ```suggestion
- "Is this change safe?"
```

> - Is it valuable?

Reviewer 4: ```suggestion
- "Is this change valuable?"
```

Simple reviews have become like linting. You shouldn't have humans be doing either.

> 12 months ago, if someone told you a line should be wrapped at a particular character, you would say "hey, I've updated the linter, you can ignore this." Similarly now, agents can spot logic problems and laziness faster, more effectively, and more consistently than a human.

Reviewer 4: ```suggestion
For example, if someone told you 12 months ago that a line should be wrapped at a particular character, you would say, "Hey, I've updated the linter, you can ignore this." Agents can basically do this for logic problems and laziness faster, more effectively, and more consistently than a human.
```

> This means we can have a world where humans review code less and the ones they do are more effective as a result. We're doing a lot of work on tools and agentic flows we can use in CI to make it safe to operate at this new, increased speed.

Reviewer 4: ```suggestion
This means we can have a world where humans review code less. And the code reviews they do need to do are more effective as a result. We're doing a lot of work on tools and agentic flows we can use in CI to make it safe to operate at this new, increased speed.
```

## What's next

> The next horizon, which I think is something that needs to be solved across the industry and we're really excited to work on, is having agents test as much of the code as possible. 

Reviewer 4: ```suggestion
The next horizon – which I think is something that needs to be solved across the industry and we're really excited to work on – is having agents test as much of the code as possible. 
```

> We're not in the old world of writing some Playwright code by hand to assess a change then running that slow test every time. Instead, we can do a subset of what a human would do: 

Reviewer 4: ```suggestion
We're not in the old world of writing some Playwright code by hand to assess a change, and then running that slow test every time. Instead, we can make an agent do a subset of what a human would do: 
```

- Click around in the application
- See if it still works
- See if it does what it claims to do

> <TeamMember name="Pawel Cebula" photo /> built an early version of this with a [`qa-frontend` skill](https://github.com/PostHog/posthog/pull/72523) that checks out a PR, design test cases, and drives the actual UI in a browser. Evidence lands as annotated screenshots, an animated reel inline, and even an optional recorded video.

Reviewer 4: ```suggestion
<TeamMember name="Pawel Cebula" photo /> built an early version of this with a [`qa-frontend` skill](https://github.com/PostHog/posthog/pull/72523) that checks out a PR, designs test cases, and drives the actual UI in a browser. The agent's output lands as annotated screenshots, an animated reel inline, and even an optional recorded video.
```

> This makes reviews easier, but also gives agents a way to escalate and say "I need a human! I don't like it, it doesn't make sense to me, it needs further thought, it isn't good enough, or doesn't work." The agent can also say "hey, I wrote a Playwright test for this because I think it should stay in the application." 

Reviewer 4: ```suggestion
This makes reviews easier, but it also gives agents a way to escalate and say "I need a human! I don't like it, it doesn't make sense to me, it needs further thought, it isn't good enough, or doesn't work, etc." The agent can also say "hey, I wrote a Playwright test for this because I think it should stay in the application." 
```

> A human could do this if they choose, but when we're going to be reviewing 2-4x as many PRs, we have to reduce the amount of work we do per PR. If we don't, we will hit a ceiling in what's possible, not because we're at the limit of the system, but because humans physically don't have the time. All our efforts here are working to avoid this.

Reviewer 4: ```suggestion
These QA skills are tasks that humans could do themselves, but the point is to reduce the amount of work we do per PR if we're reviewing 2-4x as many as before. If we don't, we'll hit a ceiling – not because we're at the limit of the system, but because humans physically don't have the time. All our efforts here are working to avoid this. That's what we believe makes for good devex in the age of agents.
```

Reviewer 4: added a conclusion-y sentence so it ends more smoothyl
